pgx/src/datum/range.rs

```diff
@@ -9,56 +9,281 @@ Use of this source code is governed by the MIT license that can be found in the
-    ptr: *mut pg_sys::varlena,
-    range_type: *mut pg_sys::RangeType,
-    _marker: PhantomData<T>,
+impl<T> RangeBound<T>
+where
+    T: RangeSubType,
+{
+    /// Returns a reference to this [`RangeBound`]'s underlying value.
+    ///
+    /// If this [`RangeBound`] is [`RangeBound::Infinite`], then `None` is returned, otherwise
+    /// `Some(&T)`
+    #[inline]
+    pub fn get(&self) -> Option<&T> {
+        match self {
+            RangeBound::Infinite => None,
+            RangeBound::Inclusive(v) | RangeBound::Exclusive(v) => Some(v),
+        }
+    }
+
+    /// Convert this pgx [`RangeBound`] into the equivalent Postgres [`pg_sys::RangeBound`].
+    ///
+    /// Note that the `lower` property is always set to false as a [`RangeBound`] doesn't know the
+    /// end on which it's placed.
+    pub fn into_pg(self) -> pg_sys::RangeBound {
+        match self {
+            RangeBound::Infinite => pg_sys::RangeBound {
+                val: pg_sys::Datum::from(0),
+                infinite: true,
+                inclusive: false,
+                lower: false,
+            },
+            RangeBound::Inclusive(v) => pg_sys::RangeBound {
+                val: v.into_datum().unwrap(),
+                infinite: false,
+                inclusive: true,
+                lower: false,
+            },
+            RangeBound::Exclusive(v) => pg_sys::RangeBound {
+                val: v.into_datum().unwrap(),
+                infinite: false,
+                inclusive: false,
+                lower: false,
+            },
+        }
+    }
+
+    /// Create a typed pgx [`RangeBound`] from an arbitrary Postgres [`pg_sys::RangeBound`].
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as it cannot guarantee that the `val` property, which is a
+    /// [`pg_sys::Datum`], points to (or is) something correct for the generic type `T`.
+    pub unsafe fn from_pg(range_bound: pg_sys::RangeBound) -> RangeBound<T> {
+        if range_bound.infinite {
+            RangeBound::Infinite
+        } else if range_bound.inclusive {
+            // SAFETY: caller has asserted that `val` is a proper Datum for `T`
+            unsafe { RangeBound::Inclusive(T::from_datum(range_bound.val, false).unwrap()) }
+        } else {
+            // SAFETY: caller has asserted that `val` is a proper Datum for `T`
+            unsafe { RangeBound::Exclusive(T::from_datum(range_bound.val, false).unwrap()) }
+        }
+    }
 }
 
-impl<T> Range<T>
+impl<T> From<&RangeBound<T>> for RangeBound<T>
 where
-    T: FromDatum + IntoDatum + RangeSubType,
+    T: RangeSubType,
 {
-    /// ## Safety
-    /// This function is safe, but requires that
-    /// - datum is not null
-    /// - datum represents a PG RangeType datum
     #[inline]
-    unsafe fn from_pg(datum: pg_sys::Datum) -> Option<Self> {
-        unsafe { Self::from_polymorphic_datum(datum, false, T::range_type_oid()) }
+    fn from(value: &RangeBound<T>) -> Self {
+        Clone::clone(value)
```

Comment:
This is an atypical impl, I'd be inclined to simply remove it.

Reply:
I think I did this to appease the compiler.  I'll double-check.

Reply:
yeah, It was kinda being used by the tests, but changing them to just `.clone()` its `RangeBound` reference is totally fine.  done.